server/src/main/java/com/linecorp/centraldogma/server/internal/mirror/MirroringMigrationService.java

```diff
@@ -0,0 +1,297 @@
+        credentialIds.add(id);
+
+        final String jsonFile = credentialFile(id);
+        project.metaRepo()
+               .commit(Revision.HEAD, System.currentTimeMillis(), Author.SYSTEM,
+                       "Migrate the credential '" + id + "' in '" + PATH_LEGACY_CREDENTIALS +
+                       "' into '" + jsonFile + "'.",
+                       Change.ofJsonUpsert(jsonFile, credential))
+               .get();
+    }
+
+    @Nullable
+    private static ArrayNode getMetaData(Project project, String path)
+            throws InterruptedException, ExecutionException {
+        final Map<String, Entry<?>> entries = project.metaRepo()
+                                                     .find(Revision.HEAD, path, ImmutableMap.of())
+                                                     .get();
+        final Entry<?> entry = entries.get(path);
+        if (entry == null) {
+            return null;
+        }
+
+        final JsonNode credentialJson = (JsonNode) entry.content();
+        if (!credentialJson.isArray()) {
+            throw new RepositoryMetadataException(
+                    path + " must be an array: " + credentialJson.getNodeType());
+        }
+        return (ArrayNode) credentialJson;
+    }
+
+    private static CommitResult rename(Project project, String oldPath, String newPath) throws Exception {
+        return project.metaRepo()
+                      .commit(Revision.HEAD, System.currentTimeMillis(), Author.SYSTEM,
+                              "Rename " + oldPath + " into " + newPath,
+                              Change.ofRename(oldPath, newPath))
+                      .get();
+    }
+
+    /**
+     * Generates a random ID for the given mirror.
+     * Pattern: {@code mirror-<projectName>-<localRepo>-<shortWord>}.
+     */
+    private String generateRandomIdForMirror(String projectName, ObjectNode mirror) {
+        if (shortWords == null) {
+            shortWords = buildShortWords();
+        }
+        final int rand = ThreadLocalRandom.current().nextInt(shortWords.size());
+        final String shortWord = shortWords.get(rand);
+        return "mirror-" + projectName + '-' + mirror.get("localRepo").asText() + '-' +
+               shortWord;
+    }
+
+    /**
+     * Generates a random ID for the given credential.
+     * Pattern: {@code credential-<projectName>-<shortWord>}.
+     */
+    private String generateRandomIdForCredential(String projectName) {
+        if (shortWords == null) {
+            shortWords = buildShortWords();
+        }
+        final int rand = ThreadLocalRandom.current().nextInt(shortWords.size());
+        final String shortWord = shortWords.get(rand);
+        return "credential-" + projectName + '-' + shortWord;
+    }
+
+    private static String uniquify(String id, Set<String> existingIds) {
+        while (existingIds.contains(id)) {
+            id += '-' + Integer.toHexString(ThreadLocalRandom.current().nextInt());
+        }
+        return id;
```

Comment:
nit; although the probability is low, it seems like the current implementation keeps adding hex strings which I don't think makes sense

```suggestion
        String tmp = id;
        while (existingIds.contains(tmp)) {
            tmp = id + '-' + Integer.toHexString(ThreadLocalRandom.current().nextInt());
        }
        return tmp;
```

Reply:
Nice catch!